pkg/api/v1beta1/types.go

```diff
@@ -98,6 +98,118 @@ type VolumeSource struct {
+	TypeMeta `json:",inline"`
+	Items []PersistentVolume `json:"items"`
+}
+
+type PersistentVolumeSpec struct {
+
+	// unique identifier of the storage device relative to the storage provider
+	// ex: an AWS EBS volume ID
+	StorageID string `json:"storageId,omitempty"`
+	Resources ResourceList `json:resources,omitempty`
+	Source VolumeSource `json:source,omitempty`
+}
+
+type PersistentVolumeStatus struct {
+	Phase StoragePhase `json:"phase,omitempty"`
+	PersistentStorageDeviceRef ObjectReference `json:storageDevice,omitempty`
+	Resources ResourceList `json:resources,omitempty`
+	Source VolumeSource `json:source,omitempty`
+}
+
+// PersistentStorageDevice is an actual storage device backed by a provider to be
+// paired with a PersistentVolume request by pod authors
+type PersistentStorageDevice struct {
+	TypeMeta   `json:",inline"`
+
+	// Spec defines the storage device
+	Spec       PersistentStorageDeviceSpec   `json:"spec,omitempty"`
+
+	// Status represents the current information about a storage device.
+	// this data may not be up to date.
+	Status     PersistentStorageDeviceStatus `json:"status,omitempty"`
+}
+
+type PersistentStorageDeviceList struct {
+	TypeMeta `json:",inline"`
+	Items []PersistentStorageDevice `json:"items"`
+}
+
+type PersistentStorageDeviceStatus struct {
+	Phase StoragePhase `json:"phase,omitempty"`
+	Source VolumeSource `json:volumeSource,omitempty`
+
+	// where the storage device backing the persistent volume is currently mounted
+	CurrentMount Mount `json:currentMount,omitempty`
+
+	PersistentVolumeRef ObjectReference `json:persistentVolumeRef,omitempty`
+}
+
+// a PersistentStorageDeviceSpec describes the common attributes of storage devices
+// and allows a Source for provider-specific attributes
+type PersistentStorageDeviceSpec struct {
+
+	// unique identifier of the storage device relative to the provider
+	// ex: an AWS EBS volume ID
+	StorageID string `json: "storageId"`
+
+	// size, iops, throughput of the storage device
+	Resources ResourceList `json:resources,omitempty`
+
+	// Source contains provider-specific information about a storage device
+	Source		VolumeSource `json:source,omitempty`
+}
+
+type Mount struct {
+	Host        string    		`json:"host,omitempty"`
+	HostIP      string    		`json:"hostIP,omitempty"`
+	MountedDate util.Time 		`json:"mountedDate,omitempty"`
+	Phase       StoragePhase 	`json:"phase,omitempty"`
+}
+
+type StoragePhase string
+
+const (
+	MountPending StoragePhase = "Pending"
+	Attached     StoragePhase = "Attached"
+	Formatting   StoragePhase = "Formatting"
+	Formatted    StoragePhase = "Formatted"
+	Mounted      StoragePhase = "Mounted"
+	MountFailed  StoragePhase = "Failed"
+	MountDelete  StoragePhase = "Deleted"
+)
+
+type AWSElasticBlockDevice struct {
+	// the device's EBS volumeID from AWS
+	VolumeID string `json:volumeId`
+}
+
+type NFSMount struct {
+	Server string `json:"server,omitempty`
+	SourcePath string `json:sourcePath,omitempty`
```

Comment:
Thanks for pointing that out.  They were in a parent struct but was accidentally dropped as we moved attributes down.  I will re-add them.